### PR TITLE
Fix bug in macro average for a number of classification metrics

### LINKED ISSUE #1664: Odd behaviour with macro averaging
## 🐛 Bug
I have a doubt concerning the macro averaging implementation.  I find that odd that given the same "pred" and "trg" tensor, if you increase the num_class parameter, the metric value decreases. It seems that the value of a class, which is not the given sample, is set to 0. Therefore, when applying “macro” averaging, this metric value set at 0 is still taken into account in the mean per class.
Here is an example with the F1 score, but this behaviour is found with other classification metrics as accuracy and recall.  

### To Reproduce

```py
import torch
import torchmetrics
from torchmetrics import F1Score
print(" Torchmetrics version {}".format(torchmetrics.__version__))
target = torch.tensor([2, 1, 0, 0])
preds = torch.tensor([2, 1, 0, 1])
for i in range(3,9):
    f1_score=F1Score(task="multiclass",num_classes=i,average="macro")
    torch_met=f1_score(preds, target)
    print("num class {} f1 score {}".format(i,torch_met))
```
Output : 
 Torchmetrics version 0.11.4
num class 3 f1 score 0.7777777910232544
num class 4 f1 score 0.5833333730697632
num class 5 f1 score 0.46666669845581055
num class 6 f1 score 0.3888888955116272
num class 7 f1 score 0.3333333432674408
num class 8 f1 score 0.2916666865348816

### Expected behavior

To my mind, we would expect the F1 score to be constant even when num_class value increases. If a class is not in a sample, it should not impact the classification metric value.
To avoid that effect, we could consider set to Nan the value of the class which are not in the sample and then apply nanmean. 
### Environment

- TorchMetrics version 0.11.4 installed with pip :
- Python 3.10 & PyTorch Version 2.0.0:

### LINKED ISSUE #1691: Accuracy is wrong when ignore_index is set
## 🐛 Bug

When the _ignore\_index_ argument is set to a value, the MulticlassAccuracy object sets the class accuracy to zero, but considers it in the average calculation.

### To Reproduce

```python
import torch
from torchmetrics.classification import MulticlassAccuracy

# simulate the output of a perfect predictor (i.e. preds == target)
target = torch.tensor([0, 1, 2, 0, 1, 2])
preds  = target

metric = MulticlassAccuracy(num_classes=3, average='none', ignore_index=0)
res = metric(preds, target)
print(res)
# it prints [0., 1., 1.]

metric = MulticlassAccuracy(num_classes=3, average='macro', ignore_index=0)
res = metric(preds, target)
print(res)
# it prints 0.6667 instead of 1
```

### Expected behavior

It should not take into account the ignored class.

### Environment

- TorchMetrics version: 0.11.4 installed from pip
- Python & PyTorch Version: 3.10.6 - 1.13.0
- Any other relevant information such as OS (e.g., Linux): Ubuntu 22.04

### LINKED ISSUE #1692: Macro Precision computation is wrong when ignore_index is used
## 🐛 Bug

When ignore_index is used in MulticlassPrecision, it zeros out the metric on the class to ignore but it still includes the value in the average computation, leading to wrong results.

### To Reproduce

```python
import torch
from torchmetrics.classification import MulticlassPrecision

# simulate the output of a perfect predictor (i.e. preds == target)
target = torch.tensor([0, 1, 2, 0, 1, 2])
preds  = target.clone()

metric = MulticlassPrecision(num_classes=3, average='none', ignore_index=0)
res = metric(preds, target)
print(res)
# it prints [0., 1., 1.]

metric = MulticlassPrecision(num_classes=3, average='macro', ignore_index=0)
res = metric(preds, target)
print(res)
# it prints 0.6667 instead of 1
```

### Expected behavior

It should not consider the ignored class in the average computation.

### Environment

- TorchMetrics version: 0.11.4 installed from pip
- Python & PyTorch Version: 3.7.9 - 1.11.0+cu113
- Any other relevant information such as OS (e.g., Linux): Linux Mint 18.3 Sylvia

### LINKED ISSUE #1693: MulticlassJaccardIndex returns 0 for classes not present in prediction and target
## 🐛 Bug

MulticlassJaccardIndex returns 0 for classes not present in prediction and target. This leads to bad (worse) results if used in conjunction with average="macro".

There was a similar issue when the metrics were still a part of pytorch lightning: https://github.com/Lightning-AI/lightning/pull/3098

### To Reproduce

MWE:

```
import torch
from torchmetrics.classification import MulticlassJaccardIndex

# edge case: class 2 is not present in the target AND the prediction
target = torch.tensor([0, 1, 0, 0])
preds = torch.tensor([0, 1, 0, 1])

metric = MulticlassJaccardIndex(num_classes=3, average="none")
print(metric(preds, target))
# tensor([0.6667, 0.5000, 0.0000])
# I would expect tensor([0.6667, 0.5000, 1.0000])

metric = MulticlassJaccardIndex(num_classes=3, average="macro")
print(metric(preds, target))
# tensor(0.3889)
```

Second example (adapted copy from https://github.com/Lightning-AI/lightning/pull/3098):
```
import torch
from torchmetrics.functional.classification.jaccard import multiclass_jaccard_index

target = torch.tensor([0, 1])
pred = torch.tensor([0, 1])
print(multiclass_jaccard_index(pred, target, num_classes=10))  # Should return tensor(1.)
print(
    multiclass_jaccard_index(pred, target, num_classes=10, average="none")
)  # Should return tensor([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])
```

### Expected behavior

See comments in the MWEs.

### Environment

- python                    3.10.9 (conda)
- pytorch                   1.13.1 (conda)
- torchmetrics              0.11.4 (pip)

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes #1693
Fixes #1692
Fixes #1691
Fixes #1664

<details>
  <summary>Before submitting</summary>

- [ ] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
@SkafteNicki how is it going here? :rabbit:

File: src/torchmetrics/functional/classification/jaccard.py
Comment (JohannesK14) [inline src/torchmetrics/functional/classification/jaccard.py:1]: Hi, I know I created https://github.com/Lightning-AI/torchmetrics/issues/1693#issue-1656343011 and suggest a IoU of 1.0 if a class is neither present in target nor prediction, but I am not that sure anymore.

Doing so quietly might lead to over optimistic values.

As I showed in https://github.com/Lightning-AI/torchmetrics/issues/1693#issuecomment-1516379416, the JaccardIndex internally updates a confusion matrix over all batches, which results in the correct values, **if a final compute() is called after the validation/testing epoch**.

**Comment by ioangatop** [comment]
Hi @SkafteNicki @Borda @JohannesK14 thank you for the fix anr / or review! I wanted to ask the state of this PR, because I also experience the issue :/

**Comment by SkafteNicki** [comment]
@ioangatop I am currently working on it locally. Nearly done, but I have some unit tests that are still failing which I need to debug. Hopefully, get it done before the end of this week and I make sure that it is done before v1.0 release.
